src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -684,29 +704,40 @@ public override void VisitFunctionCallSyntax(FunctionCallSyntax syntax)
 
                 switch (bindings[syntax])
                 {
-                    case UnassignableSymbol errorSymbol:
+                    case ErrorSymbol errorSymbol:
                         // function bind failure - pass the error along
-                        return UnassignableTypeSymbol.CreateErrors(errors.Concat(errorSymbol.GetDiagnostics()));
+                        return ErrorType.Create(errors.Concat(errorSymbol.GetDiagnostics()));
 
                     case FunctionSymbol function:
                         return GetFunctionSymbolType(function, syntax.OpenParen, syntax.CloseParen, syntax.Arguments, argumentTypes, errors);
 
                     default:
-                        return UnassignableTypeSymbol.CreateErrors(errors.Append(DiagnosticBuilder.ForPosition(syntax.Name.Span).SymbolicNameIsNotAFunction(syntax.Name.IdentifierName)));
+                        return ErrorType.Create(errors.Append(DiagnosticBuilder.ForPosition(syntax.Name.Span).SymbolicNameIsNotAFunction(syntax.Name.IdentifierName)));
                 }
             });
 
         public override void VisitInstanceFunctionCallSyntax(InstanceFunctionCallSyntax syntax)
             => AssignType(syntax, () => {
-
                 var errors = new List<ErrorDiagnostic>();
-                
+
                 var baseType = typeManager.GetTypeInfo(syntax.BaseExpression);
                 CollectErrors(errors, baseType);
 
-                if (errors.Any())
+                if (PropagateErrorType(errors, baseType))
                 {
-                    return UnassignableTypeSymbol.CreateErrors(errors);
+                    return ErrorType.Create(errors);
+                }
+
+                if (baseType is ResourceType resourceType)
+                {
+                    // We're accessing a property on the resource body.
+                    baseType = resourceType.Body.Type;
+                }
+
+                if (!(baseType is ObjectType objectType))

```

Comment:
Discriminated objects aren't derived from ObjectType. Is this going to break in that case?

Reply:
Methods and discriminators are two independent things, so it's weird to only support methods with non-discriminated objects. It's ok if it goes in a separate PR, but we should do it sooner rather than later.

Reply:
Btw, can you create an issue to track this. We're going to forget about this otherwise and we'll be super surprised when we discover a discriminated object with list* functions.

Reply:
Done: https://github.com/Azure/bicep/issues/764